feat: Add "Watch" tab to the debug panel

tonyanziano opened the pull request:
## Description

This PR adds the **Watch** tab to the debugging panel.

The Watch tab allows users to observe arbitrary variables stored in the bot's different memory scopes while interacting with the bot.

## Task Item

Fixes #7978

## Screenshots

**Bot not yet started**

![image](https://user-images.githubusercontent.com/3452012/123470051-b69e9400-d5a8-11eb-81d3-9548ed46331c.png)


**Talking to bot**

![image](https://user-images.githubusercontent.com/3452012/123470139-d7ff8000-d5a8-11eb-9af3-56aa227d1b02.png)

![image](https://user-images.githubusercontent.com/3452012/123470278-0b420f00-d5a9-11eb-979e-03d1a67d6770.png)


**Validation**

![image](https://user-images.githubusercontent.com/3452012/123470177-e64d9c00-d5a8-11eb-8500-e76668189234.png)

hatpick commented on Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WatchTab/WatchTabContent.tsx:
```diff
@@ -0,0 +1,326 @@
+  iconName: 'Cancel',
+};
+
+const NameColumnKey = 'column1';
```
use a more descriptive name? :D

hatpick commented on Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WatchTab/WatchTabContent.tsx:
```diff
@@ -0,0 +1,326 @@
+        return undefined;
+      }
+    },
+    undefined
```
?

tonyanziano replied:
This is basically the equivalent of:

```
const segments = ['user.outer.inner.someValue'].split('.')
return botTrace?.value?.[segments[0]]?.[segments[1]]?.[segments[2]]?....[segments[n]]
```

Do you know of a better way to do this with a dynamic number of properties to try and access? This basically works its way from the root of the object down to the desired property.

tonyanziano replied:
I suppose I could do something like:

```
let returnVal = botTrace?.value;
for const segment in segments.split('.') {
  returnVal = returnVal?.[segment]
}
return returnVal
```

tonyanziano replied:
^^ Just tried this, it works and looks way cleaner.

srinaath replied:
Sweet. I had a similar requirement as well. Used lodash get
https://github.com/microsoft/BotFramework-Composer/pull/8141/files#diff-c949aaa883b99af0c85135f197dfbdf385d73097d1f5ba2511fbe540a4f39a49R60.

tonyanziano replied:
Awesome, I'll do that instead.

hatpick commented on Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WatchTab/WatchTabContent.tsx:
```diff
@@ -0,0 +1,326 @@
+  return value;
+};
+
+export const WatchTabContent: React.FC<DebugPanelTabHeaderProps> = ({ isActive }) => {
```
why is the tab knows it's active or not, isn't the parent deciding to show it or not?

tonyanziano replied:
This is the pattern that the debug panel tabs use so I am following it. I do not know the reasoning behind the tabs knowing if they are active or not, but I do not think rewriting it should be part of this PR.

srinaath replied:
The active property is sent to all the tabs but only 1 uses it and is controlled at the parent level. The reason for isActive property passed through to one of the tabs is for implementing the unread notification indicator. 
https://github.com/microsoft/BotFramework-Composer/blob/main/Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WebChatLog/WebChatLogItemHeader.tsx#L23

hatpick commented on Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WatchTab/WatchTabContent.tsx:
```diff
@@ -0,0 +1,326 @@
+    data: { properties: [] },
+  });
+
+  const watchedVarsSelection = useMemo(
```
this is practically a ref

hatpick commented on Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WatchTab/WatchTabContent.tsx:
```diff
@@ -0,0 +1,326 @@
+  const onSelectPath = useCallback(
+    (variableId: string, path: string) => {
+      const watchedVar = Object.values(watchedVars).find((varPath) => varPath === path);
+      if (watchedVar) {
```
have we considered a UX in which we dont show the selected vars in the list OR we disable them?

tonyanziano replied:
1. What would be a scenario in which we don't show the selected variables in the list?

2. There are no designs at the moment for disabled variables

hatpick replied:
when a variable is selected and they re-select it, it shows an error, which I think is too much, I think it should be silent about it, or if we consider it an "error" we can easily prevent it by not showing the selected variables in the select list or show them as disabled with a selected UI treatment (an explicit label or color etc.)

tonyanziano replied:
Ahh I see. It might be the wording / naming of the code that is confusing you here.

"onSelectPath" is actually the "Submit" action of the picker. So let's say you have a row in the table that is already watching `user.myValue`. If you then add another row in the table, type in `user.myValue` and hit "Enter" you will see this validation error underneath the duplicate row and the duplicate **will not** be watched.

I spoke to @mewa1024 about this and we felt that this was a better UX than silently failing or doing a no-op.

tonyanziano replied:
Here's an example:

![image](https://user-images.githubusercontent.com/3452012/122836097-78525d80-d2a6-11eb-8a56-7cb49e1be9a0.png)

hatpick replied:
@mewa1024 @tonyanziano thanks for explaining the use case, it wasn't far from what I imagined. 👍🏼 

Chrome/edge dev tools have no issues with devs adding the same variable multiple times, I thought we may wanna consider following that.

tonyanziano replied:
Ahhh, interesting! I wonder what VS / VS Code do in this situation.

tonyanziano replied:
VS Code lets you do this without any feedback at all:

![image](https://user-images.githubusercontent.com/3452012/122838757-97072300-d2ab-11eb-9599-f02d36567520.png)

hatpick replied:
another point I wanted to mention is captured in this screenshot above, "not available", if user enters something random that doesn't exist or they think it does but it doesn't, "not available" is a great indicator vs. undefined (if that's the case)
@mewa1024

mewa1024 replied:
agree with all of the above. some other context about the red error message: I think Tony and I landed on that instead of just hiding the variable from the list because the user could type in something freeform (since it's a combobox-ish input rather than a pure dropdown input). 

About the "not available" versus "undefined", "not available" sounds good to me (I'll update the mocks for future reference.)

hatpick replied:
@mewa1024 just to clarify, if a variable is indeed available but has no value at watch time, it would be undefined, but if it's not available (doesn't exist in the list, like user thinks it's there but it's not) it shows unavailable.

mewa1024 replied:
ah, thanks for the clarification. in the JSON for the bot, would an undefined thing look like this:

`"propertyname" : "undefined"`

if so, seems like it should be shown in monospaced font, like this:
![image](https://user-images.githubusercontent.com/39317542/122969796-ed26a580-d341-11eb-9fdf-555481dfd8ac.png)

rather than this:
![image](https://user-images.githubusercontent.com/39317542/122969757-dc762f80-d341-11eb-8d4c-f34be59f1c79.png)

Thoughts?

tonyanziano replied:
I think our editor only allows you to set a value to `null` if it is not set, if it is actually `undefined`, then I believe Monaco just omits the value from the display:

So here's what it would look like:

![image](https://user-images.githubusercontent.com/3452012/123307238-1c225000-d4d7-11eb-9b1b-7e6aa648b5d0.png)


If we had a `prop4: undefined` it would not even show up.

hatpick commented on Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WatchTab/WatchTabContent.tsx:
```diff
@@ -0,0 +1,326 @@
+
+  // we need to refresh the details list every time a new bot state comes in
+  const refreshedWatchedVars = useMemo(() => {
+    return Object.entries(watchedVars).map(([key, value]) => {
```
use {...watchedVars} ?

tonyanziano replied:
it needs to be of type `{ key, value }[]` so this won't work

hatpick commented on Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WatchTab/WatchTabContent.tsx:
```diff
@@ -0,0 +1,326 @@
+  }, [mostRecentBotState, pickerErrorMessages, watchedVars]);
+
+  const renderRow = useCallback((props?: IDetailsRowProps) => {
+    if (props) {
```
nit: condition ? this : that

hatpick commented on Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WatchTab/WatchTabContent.tsx:
```diff
@@ -0,0 +1,326 @@
+    return null;
+  }, []);
+
+  const renderColumn = useCallback(
```
I think it's much better if you add each column renderer to the column itself.

hatpick commented on Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WatchTab/WatchTabContent.tsx:
```diff
@@ -0,0 +1,326 @@
+              return <WatchTabObjectValue value={value} />;
+            } else if (value === undefined) {
+              // render undefined indicator
+              return <span css={undefinedValue}>undefined</span>;
```
formatMessage?

hatpick commented on Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WatchTab/WatchTabContent.tsx:
```diff
@@ -0,0 +1,326 @@
+  }, [watchedVars]);
+
+  const onClickRemove = () => {
+    const updated = produce(watchedVars, (draftState) => {
```
this is an overkill to bring in immer here, just delete and spread maybe instead?

hatpick commented on Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WatchTab/WatchTabContent.tsx:
```diff
@@ -0,0 +1,326 @@
+  };
+
+  const removeIsDisabled = useMemo(() => {
+    return selectedVars === undefined || selectedVars.length === 0;
```
!selectedVars?.length

hatpick commented on Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WatchTab/config.ts:
```diff
@@ -0,0 +1,16 @@
+
+export const WatchTabConfig: TabExtensionConfig = {
+  key: WatchTabKey,
+  description: () => formatMessage('Watch tab'),
```
this wont work when locale changes, convert the whole thing to a function

tonyanziano replied:
I'm not familiar with how the strings for each locale are generated. What needs to be converted to a function here?

hatpick commented on Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WatchTab/index.ts:
```diff
@@ -0,0 +1,4 @@
+// Copyright (c) Microsoft Corporation.
```
avoid barrel files for the sake of webpack tree shaking

hatpick commented on Composer/packages/client/src/pages/design/DebugPanel/WatchVariableSelector/WatchVariablePicker.tsx:
```diff
@@ -0,0 +1,251 @@
+const textFieldStyles = (errorMessage?: string): Partial<ITextFieldStyles> => ({
+  field: {
+    fontFamily: DEFAULT_FONT_SETTINGS.fontFamily,
+    fontSize: 12,
```
use fluenttheme font size

hatpick commented on Composer/packages/client/src/pages/design/DebugPanel/WatchVariableSelector/WatchVariablePicker.tsx:
```diff
@@ -0,0 +1,251 @@
+export const WatchVariablePicker = React.memo((props: WatchVariablePickerProps) => {
+  const { errorMessage, payload, variableId, path, onSelectPath } = props;
+  const [query, setQuery] = useState(path);
+  const inputBoxElement = useRef<ITextField | null>(null);
```
inputBoxElementRef makes it clear that it's a ref

hatpick commented on Composer/packages/client/src/pages/design/DebugPanel/WatchVariableSelector/WatchVariablePicker.tsx:
```diff
@@ -0,0 +1,251 @@
+  const { errorMessage, payload, variableId, path, onSelectPath } = props;
+  const [query, setQuery] = useState(path);
+  const inputBoxElement = useRef<ITextField | null>(null);
+  const pickerContainerElement = useRef<null | HTMLDivElement>(null);
```
same here

hatpick commented on Composer/packages/client/src/pages/design/DebugPanel/WatchVariableSelector/WatchVariablePicker.tsx:
```diff
@@ -0,0 +1,251 @@
+
+  useEffect(() => {
+    handleDebouncedSearch();
+  }, [menuItems, flatPropertyListItems, noSearchResultMenuItem, query]);
```
handleDebouncedSearch as dep should be enough

hatpick commented on Composer/packages/client/src/pages/design/DebugPanel/WatchVariableSelector/WatchVariablePicker.tsx:
```diff
@@ -0,0 +1,251 @@
+  const onTextBoxKeyDown = useCallback(
+    (event: KeyboardEvent<HTMLInputElement>) => {
+      // enter
+      if (event.keyCode == 13) {
```
event.key === 'Enter', keyCodes are deprecated

hatpick commented on Composer/packages/client/src/pages/design/DebugPanel/WatchVariableSelector/WatchVariablePicker.tsx:
```diff
@@ -0,0 +1,251 @@
+            width: '240px',
+          },
+        }}
+        target={pickerContainerElement}
```
pickerContainerElement.current

hatpick commented on Composer/packages/client/src/pages/design/DebugPanel/WatchVariableSelector/utils/components/PickerContextualMenuItem.tsx:
```diff
@@ -0,0 +1,67 @@
+  root: { whiteSpace: 'nowrap', overflow: 'hidden', textOverflow: 'ellipsis', height: defaultTreeItemHeight },
+};
+
+export const GetPickerContextualMenuItem = (query: string, propertyTreeExpanded: Record<string, boolean>) => (
```
camelCase naming: getPickerContextualMenuItem

srinaath commented on Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WatchTab/WatchTabContent.tsx:
```diff
@@ -0,0 +1,326 @@
+  const rawWebChatTraffic = useRecoilValue(webChatTrafficState(currentProjectId ?? ''));
+  const [watchedVars, setWatchedVars] = useState<Record<string, string>>({});
+  const [pickerErrorMessages, setPickerErrorMessages] = useState<Record<string, string>>({});
+  const [selectedVars, setSelectedVars] = useState<IObjectWithKey[]>();
```
Can we use the expanded version itself? 
selectedVariables
watchedVariables

srinaath commented on Composer/packages/client/src/recoilModel/dispatchers/webchat.ts:
```diff
@@ -5,14 +5,20 @@
     const { set } = callbackHelpers;
     set(webChatTrafficState(projectId), []);
     set(webChatInspectionDataState(projectId), undefined); // clear the inspection panel
+    set(watchedVariablesState(projectId), []); // TODO: might not want to do this depending on how annoying it is for the user -- do you want to wipe variables when you restart convo?
```
@tonyanziano We definitely dont want to wipe out the variable list on bot restart like you said following the VSCode model

srinaath commented on Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WatchTab/WatchTabContent.tsx:
```diff
@@ -0,0 +1,326 @@
+          // render the value display
+          if (mostRecentBotState) {
+            const value = getValueFromBotTraceScope(item.value, mostRecentBotState?.activity);
+            if (value !== null && typeof value === 'object') {
```
how are we rendering the value `null`

mewa1024 replied:
@sorgh and I chatted about this offline and here is an example of a table that is watching variables that return an object, "undefined", "null", or are not available:
![image](https://user-images.githubusercontent.com/39317542/122978338-42b38000-d34b-11eb-8ac9-8c3bba30b7c5.png)

"unavailable" in Segoe gray 130 (#605E5C); "undefined" and "null"  appear the same as the other values (same monospace font families and color coding).

Will this work?

srinaath replied:
That should work for us. Also, unrelated, boolean would just be rendered as true or false without the string interpolation "true".

beyackle commented on Composer/packages/client/src/pages/design/DebugPanel/WatchVariablePicker/WatchVariablePicker.tsx:
```diff
@@ -0,0 +1,263 @@
+
+        const predicate = getFilterPredicate(query);
+
+        const filteredItems = searchableItems.filter(predicate);
```
`.filter()` makes a new array anyway, so there's no need to reassign flatPropertyListItems to searchableItems in here. I think just `const filteredItems = flatPropertyListItems.filter(predicate)` ought to work.

hatpick commented on Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WatchTab/WatchTabContent.tsx:
```diff
@@ -0,0 +1,325 @@
+    maxHeight: `calc(100% - ${toolbarHeight}px)`,
+  },
+  contentWrapper: {
+    overflowY: 'auto' as 'auto',
```
nit: as ```any```

hatpick commented on Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WatchTab/WatchTabContent.tsx:
```diff
@@ -0,0 +1,325 @@
+          // error can be due to abort
+        }
+      })();
+    }
```
effect should return the following:
```
return () => {
  abortController.abort();
};
```

hatpick commented on Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WatchTab/WatchTabContent.tsx:
```diff
@@ -0,0 +1,325 @@
+
+  // TODO: update to office-ui-fabric-react@7.170.x to gain access to "flexGrow" column property to distribute proprotional column widths
+  // (name column takes up 1/3 of space and value column takes up the remaining 2/3)
+  const watchTableColumns: IColumn[] = [
```
nit: this can be memoized

hatpick commented on Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WatchTab/WatchTabContent.tsx:
```diff
@@ -0,0 +1,325 @@
+    return !selectedVariables?.length;
+  }, [selectedVariables]);
+
+  function onRenderDetailsHeader(props, defaultRender) {
```
be consistent with arrow vs. traditional function definition

hatpick commented on Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WatchTab/WatchTabContent.tsx:
```diff
@@ -0,0 +1,325 @@
+
+  return (
+    <Stack verticalFill>
+      <Stack.Item
```
use CommandBar? (it has accessibility + points vs. current implementation)

hatpick commented on Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WatchTab/WatchTabObjectValue.tsx:
```diff
@@ -0,0 +1,58 @@
+`;
+
+type WatchTabObjectValueProps = {
+  value: object;
```
any is probably a better choice

hatpick commented on Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WatchTab/WatchTabObjectValue.tsx:
```diff
@@ -0,0 +1,58 @@
+        editorSettings={{
+          fadedWhenReadOnly: false,
+          fontSettings: {
+            fontFamily: DEFAULT_FONT_SETTINGS.fontFamily,
```
should this be using font settings from user settings or default?

tonyanziano replied:
this should probably use the user settings

hatpick commented on Composer/packages/client/src/pages/design/DebugPanel/WatchVariablePicker/WatchVariablePicker.tsx:
```diff
@@ -0,0 +1,261 @@
+
+const textFieldStyles = (errorMessage?: string): Partial<ITextFieldStyles> => ({
+  field: {
+    fontFamily: DEFAULT_FONT_SETTINGS.fontFamily,
```
same comment

tonyanziano replied:
If the user settings font affects the form editors (all inputs in the app) then it should probably use the user settings font, otherwise I think it should use the default because it is a text input and not a code editor

hatpick commented on Composer/packages/client/src/pages/design/DebugPanel/WatchVariablePicker/WatchVariablePicker.tsx:
```diff
@@ -0,0 +1,261 @@
+  const onDismiss = useCallback(() => {
+    setPropertyTreeExpanded({});
+    onHideContextualMenu();
+  }, []);
```
onHideContextualMenu to deps

hatpick commented on Composer/packages/client/src/pages/design/DebugPanel/WatchVariablePicker/WatchVariablePicker.tsx:
```diff
@@ -0,0 +1,261 @@
+  useEffect(() => setItems(menuItems), [menuItems]);
+
+  const handleDebouncedSearch: () => void = useCallback(
+    debounce(() => {
```
does this work? The debounced function gets redefined everytime any of the deps including query changes.

tonyanziano replied:
looking into it

hatpick commented on Composer/packages/client/src/pages/design/DebugPanel/WatchVariablePicker/WatchVariablePicker.tsx:
```diff
@@ -0,0 +1,261 @@
+  );
+
+  useEffect(() => {
+    handleDebouncedSearch();
```
adding to the comment above, this effect is pretty confusing

hatpick commented on Composer/packages/client/src/pages/design/DebugPanel/WatchVariablePicker/utils/components/PickerContextualMenuItem.tsx:
```diff
@@ -0,0 +1,67 @@
+
+import { PropertyItem, PropertyTreeItem } from './PropertyTreeItem';
+
+const defaultTreeItemHeight = 36;
```
this is redefined in the next file below too, maybe reuse from a constants.ts?

srinaath reviewed (dismissed):
@tonyanziano Im running the branch with its latest commits. The following are the issues im seeing.
1. The "Watch Variables" seems to be persisted across bots. The watch variables should be a bot state should wipe out when opening a new bot
2. ~~Collapsing and opening the Debug Panel  seems to wipe off variables.~~ FIXED
3. New variables added to the memory scope are showing up in the "Send a Response" variable picker but not in the variable picker inside the "Watch" Tab. Im assuming some sort of memoization issue.
4. ~~Boolean variables are showing up as a string. For ex "user.greeted".
Let me know if you are seeing the same bugs as well.~~ -- false alarm

tonyanziano commented:
I have addressed all of Soroush's comments and also adjusted some styles that looked off from the specs. Here is the updated tab with new "not available" message and placeholder text:

![watch-tab-empty](https://user-images.githubusercontent.com/3452012/123708897-07b8bd00-d821-11eb-8535-0da3c08e4082.PNG)

@srinaath I still can't repro bug 4 as seen in the screenshot, but I can confirm bugs 1-3 and am working to fix these.

srinaath commented:
@tonyanziano Ignore bug 4. I just realized a bot logic error which was converting the boolean to string to on the variable. So basically it was casting to "true". So we are good on that front.
![Screen Shot 2021-06-28 at 3 29 02 PM](https://user-images.githubusercontent.com/13004779/123711650-aa733a80-d825-11eb-9230-a67a5f1c1fd8.png)

srinaath commented on Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WatchTab/WatchTabObjectValue.tsx:
```diff
@@ -0,0 +1,54 @@
+`;
+
+const objectCellStyle = (numLinesOfJson: number) => css`
+  height: ${numLinesOfJson * 18}px;
```
@tonyanziano On a huge JSON like the dialog object we would need to cap the height of the cell. Basically, allowing scroll to inside the editor. Can we confirm the max height of the value cell @mewa1024 
![Screen Shot 2021-06-28 at 3 47 06 PM](https://user-images.githubusercontent.com/13004779/123712918-28384580-d828-11eb-8b29-df2abcea19eb.png)

tonyanziano replied:
Great point!

mewa1024 replied:
What about 80px max height? That would show 4 full lines, plus the top of a 5th line, if any. (Are we assuming that if it's an enormous JSON object the user would copy and paste into a code editor to view the whole thing?)

srinaath replied:
Yep.

srinaath commented on Composer/packages/client/src/pages/design/DebugPanel/TabExtensions/WatchTab/WatchTabContent.tsx:
```diff
@@ -0,0 +1,336 @@
+  valuePath: string,
+  botTrace: Activity
+): { value: any; propertyIsAvailable: boolean } => {
+  const pathSegments = valuePath.split('.');
```
When we are trying to access memory scope properties. For ex `dialog` or `conversation` as whole. The valuePath.split('.') just leads to one element and gets popped up. If the goal is to check if the path exists or not can we just not use 
```javascript
const valueOfProp = get(botTrace?.value, valuePath, undefined)
if(valueOfProp === undefined) then the propertyIsAvailable is false?
```

tonyanziano replied:
I'm using `propertyIsAvailable` to distinguish between an `undefined` value and an `unavailable` value because the design has different treatments for each.

`undefined` meaning that the property could exist on the parent key, but it is not set
`unavailable` meaning that the property **could not and does not exist** because the property's parent path is not set

That being said, as far as I know, we do not offer a way for users to set a property to undefined in the form editor, because `"someProperty": undefined` is not actually valid JSON even though it is valid JavaScript.

If we treat `undefined` values as `unavailable` then we don't need both treatments, but that might be confusing to users because standard dev tools treat them differently:

![image](https://user-images.githubusercontent.com/3452012/123714134-9b42bb80-d82a-11eb-98f1-b90db51f6f25.png)

So I guess this all depends on what the design is: Do we want to call all undefined / unset values `unavailable`?

@mewa1024

mewa1024 replied:
yes, the design was trying to match what users would expect from "Watch" in dev tools, namely, different treatments for "not available" and "undefined". (@sorgh, any opinions? I think you mentioned the not available/undefined difference earlier.)

tonyanziano replied:
related but separate:

to Srinaath's point, there is actually a bug when trying to watch the root memory scopes themselves `user`, `conversation`, etc. and I am trying to fix that

hatpick replied:
> yes, the design was trying to match what users would expect from "Watch" in dev tools, namely, different treatments for "not available" and "undefined". (@sorgh, any opinions? I think you mentioned the not available/undefined difference earlier.)

undefined is an accepted value, not available is for the gibberish user enters freely in the picker like:
user.name is defined by the user and until it's set, it's undefined
soroush.is.awesome is not available but it's true

tonyanziano replied:
so would the distinction then be:

`undefined` is anything not set in a **defined memory scope** (user, conversation, customUserScope, etc.)
`not available` is anything outside one of these scopes

tonyanziano commented:
All of the bugs / issues have been fixed and styles updated to appear more like the updated designs:

![watch-tab-input-padding](https://user-images.githubusercontent.com/3452012/123874666-5f245f00-d8ed-11eb-9bbb-c10a04df5e4b.PNG)
![watch-tab-values](https://user-images.githubusercontent.com/3452012/123874668-5fbcf580-d8ed-11eb-8510-a5d27dc80c90.PNG)
![watch-tab-empty-state](https://user-images.githubusercontent.com/3452012/123874669-5fbcf580-d8ed-11eb-8bdb-8759fae0e2d3.PNG)

mewa1024 commented:
@tonyanziano minor thing, can the table be slightly narrower so there is 16px of space on either side? where the pink lines are:
![image](https://user-images.githubusercontent.com/39317542/123994497-47e38100-d982-11eb-8b73-300277a3b181.png)

tonyanziano commented:
@mewa1024 Fixed spacing:

![table-padding-1](https://user-images.githubusercontent.com/3452012/124006303-ed045680-d98e-11eb-8de7-3a402e828317.PNG)

![table-padding-2](https://user-images.githubusercontent.com/3452012/124006311-ef66b080-d98e-11eb-951d-8e4f20499491.PNG)
